sympy/integrals/laplace.py

```diff
@@ -1089,6 +1092,211 @@ def simp_exp(arg):
     return _simplify(f.subs(t, t_), simplify), cond
 
 
+def _complete_the_square_in_denom(f, s):
+    from sympy.simplify.radsimp import fraction
+    [n, d] = fraction(f)
+    if d.is_polynomial(s):
+        cf = d.as_poly(s).all_coeffs()
+        if len(cf)==3:
+            a, b, c = cf
+            d = a*((s+b/(2*a))**2+c/a-(b/(2*a))**2)
+    return n/d
+
+
+def _inverse_laplace_build_rules(s, t):
+    """
+    This is an internal helper function that returns the table of inverse
+    Laplace transform rules in terms of the time variable `t` and the
+    frequency variable `s`.  It is used by `_inverse_laplace_apply_rules`.
+    """
+    a = Wild('a', exclude=[s])
+    b = Wild('b', exclude=[s])
+    c = Wild('c', exclude=[s])
+    d = Wild('d', exclude=[s])
+    global _ILT_vars
+    global _ILT_rules
```

Comment:
I'd think that global variables are evil and avoid them whenever possible.
Is it not possible to localize use `@cacheit`?

Reply:
The rules are not immutable, the time variable can change, so I don't think `@cacheit` will do the trick.

What exactly do you mean by alpha conversion?

Reply:
What I mean by alpha conversion is that you can need to rewrite to used in the pattern

For example, if this is the pattern
```
s = Dummy('s')
t = Dummy('t')
a = Wild('a', exclude=[t])
pattern = exp(-a / t) / sqrt(t)
rewrite = sqrt(pi/s)*exp(-2*sqrt(a*s))
```
And if this is the expression to match:
```
x = Symbol('x')
y = Symbol('y')
c = Symbol('c')
expr = exp(-c / x) / sqrt(x)
```
Even if the expression doesn't contain exact $s, t$

You can use this to substitute twice to match the pattern
```
match = expr.subs({x: t, y: s}).match(pattern) 
rewrite.subs(match).subs({t: x, s: y})
```
This should allow building the table only once with one set of symbols

Reply:
> This should allow building the table only once with one set of symbols

What would you build the table with?  Global dummies?

Reply:
I had done this trick
```
-def _laplace_build_rules(t, s):
+def _laplace_build_rules():
     """
     This is an internal helper function that returns the table of Laplace
     transform rules in terms of the time variable `t` and the frequency
@@ -287,6 +287,9 @@ def _laplace_build_rules(t, s):
     to the expression before matching. For most rules it should be
     ``_laplace_deep_collect``.
     """
+    s = Dummy('s')
+    t = Dummy('t')
+
     a = Wild('a', exclude=[t])
     b = Wild('b', exclude=[t])
     n = Wild('n', exclude=[t])
@@ -467,7 +470,7 @@ def _laplace_build_rules(t, s):
     (besselk(0, a*t), log((s + sqrt(s**2-a**2))/a)/(sqrt(s**2-a**2)),
      S.true, -re(a), dco) # 4.16.23
     ]
-    return laplace_transform_rules
+    return laplace_transform_rules, t, s
 
 
 def _laplace_rule_timescale(f, t, s):
@@ -759,13 +762,14 @@ def _laplace_apply_simple_rules(f, t, s):
     This function applies all simple rules and returns the result if one
     of them gives a result.
     """
-    simple_rules = _laplace_build_rules(t, s)
+    simple_rules, _t, _s = _laplace_build_rules()
     prep_old = ''
     prep_f = ''
     for t_dom, s_dom, check, plane, prep in simple_rules:
         if not prep_old==prep:
-            prep_f = prep(f)
+            prep_f = prep(f.subs({s: _s, t: _t}))
             prep_old = prep
+
         ma = prep_f.match(t_dom)
         if ma:
             try:
@@ -779,7 +783,12 @@ def _laplace_apply_simple_rules(f, t, s):
                 debug('      f:     %s'%(f,))
                 debug('      rule:  %s o---o %s'%(t_dom, s_dom))
                 debug('      match: %s'%(ma, ))
-                return (s_dom.xreplace(ma), plane.xreplace(ma), c)
+                print('hello')
+                return (
+                    s_dom.xreplace(ma).xreplace({_s: s, _t: t}),
+                    plane.xreplace(ma).xreplace({_s: s, _t: t}),
+                    c
+                )
     return None
```

Reply:
Ah! I finally understand it, thanks a lot!